### PR TITLE
✨ Router scroll behaviour, fix #1071

### LINKED ISSUE #1071: Add native-like scroll behaviour when changing routes
## Feature request

Currently the user stays at his current scroll position when the route is being changed. But when the user gets to a new page, he should see the top of the new page and when he goes back- and forward he should see the scroll position at which he left the page previously.

The logic on how the scroll position changes based on a route change should look like this:

* New page → top scroll position
* Already visited page → scroll position at which user previously left the page
* Anchor change → scroll position at anchor

#### What problem does this feature solve?

* The content of a page is ordered from top to bottom, so the top is what the user should see first when navigating to a page.
* The user expects to be at the top of a page when the route changes, everything else can lead to confusion.

#### What does the proposed API look like?

The VuePress API doesn't change with this feature.

#### How should this be implemented in your opinion?

Vue Router provides a [scrollBehavior API](https://router.vuejs.org/guide/advanced/scroll-behavior.html) which can set the scroll position when changing routes. This could be implemented into the router initialisation in `packages/@vuepress/core/lib/app/app.js`.

The anchor scrolling doesn't need to be implemented as this is already solved, but maybe it's better to put everything into one place.

#### Are you willing to work on this yourself?

Yes, just want to check if this is desired.

### PR DESCRIPTION
**Pull request by dcastil**
**Summary**

closes #1071 

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI of default theme, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature
  → see #1071 
- [ ] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

I removed `smooth-scroll` as it was causing issues when used with `active-header-links`.

### CODE REVIEW

File: packages/@vuepress/core/lib/app/app.js [hunk 1]
```diff
@@ -62,15 +62,25 @@ export function createApp (isServer) {
     base: siteData.base,
     mode: 'history',
     fallback: false,
-    routes
+    routes,
+    scrollBehavior (to, from, savedPosition) {
```
Comment (janvennemann) [inline packages/@vuepress/core/lib/app/app.js]: What about cases when `to.hash` is set?

File: packages/@vuepress/core/lib/app/app.js [hunk 2]
```diff
@@ -62,15 +62,25 @@ export function createApp (isServer) {
...
+    }
   })
 
   // redirect /foo to /foo/
   router.beforeEach((to, from, next) => {
     if (!/(\/|\.html)$/.test(to.path)) {
-      next(Object.assign({}, to, {
-        path: to.path + '/'
-      }))
+      next(
```
Comment (ulivz) [inline packages/@vuepress/core/lib/app/app.js]: Do not format code and only focus what you are doing.

File: packages/@vuepress/core/lib/app/app.js [hunk 3]
```diff
@@ -94,7 +104,11 @@ export function createApp (isServer) {
       render (h) {
         return h('div', { attrs: { id: 'app' }}, [
           h('router-view', { ref: 'layout' }),
-          h('div', { class: 'global-ui' }, globalUIComponents.map(component => h(component)))
+          h(
```
Comment (ulivz) [inline packages/@vuepress/core/lib/app/app.js]: Same here.

File: packages/@vuepress/core/lib/app/app.js (hunk 1 above)
Comment (dcastil) [reply packages/@vuepress/core/lib/app/app.js]: That interferes with `plugin-active-header-links` unfortunately. It causes the website to jump around when the anchor is being pushed.

File: packages/@vuepress/core/lib/app/app.js (hunk 2 above)
Comment (dcastil) [reply packages/@vuepress/core/lib/app/app.js]: Ah sorry, I'm using prettier by default.

File: packages/@vuepress/core/lib/app/app.js (hunk 1 above)
Comment (janvennemann) [reply packages/@vuepress/core/lib/app/app.js]: Alright perfect, then i guess in the old implementation it was just to disable scroll behavior via some flag.

Comment (dcastil) [reply packages/@vuepress/core/lib/app/app.js]: Ah that would make sense here, too! Have to check that out later.

**Pull request merged**